calabash/Classes/FranklyServer/Routes/LPTypeStringRoute.m

```diff
@@ -0,0 +1,543 @@
+#warning This file must be compiled with ARC. Use -fobjc-arc flag (or convert project to ARC).
+#endif
+
+#import "LPTypeStringRoute.h"
+#import "LPJSONUtils.h"
+#import "LPInvoker.h"
+#import "LPInvocationResult.h"
+#import "UIScriptParser.h"
+#import "LPCocoaLumberjack.h"
+#import "LPTouchUtils.h"
+
+static NSString const *LPTypeStringFirstResponderQuery = @"* isFirstResponder:1 index:0";
+
+static CFTimeInterval const LPTypeStringPostKeyDelay = 0.1;
+
+@interface LPTypeStringRoute ()
+
+@property(assign) CFTimeInterval postKeyDelay;
+
+@end
+
+@implementation LPTypeStringRoute
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _postKeyDelay = LPTypeStringPostKeyDelay;
+  }
+  return self;
+}
+
+- (BOOL) supportsMethod:(NSString *) method atPath:(NSString *) path {
+  return [method isEqualToString:@"POST"];
+}
+
+- (NSError *)errorWithDescription:(NSString *)description {
+  return [NSError errorWithDomain:@"sh.calaba.LPServer"
+                             code:1
+                         userInfo:@{ NSLocalizedDescriptionKey : description }];
+}
+
+- (NSDictionary *)failureResponseWithReason:(NSString *)reason {
+  return @{
+      @"outcome" : @"FAILURE",
+      @"reason" : reason,
+      @"details" : @""
+  };
+}
+
+- (NSDictionary *)successResponseWithResult:(id)result {
+  return @{
+      @"outcome" : @"SUCCESS",
+      @"results" : @[result]
+  };
+}
+
+- (id)firstResponder {
+  NSString *query = @"* isFirstResponder:1 index:0";
+  UIScriptParser *parser = [[UIScriptParser alloc] initWithUIScript:query];
+  [parser parse];
+  NSArray *allWindows = [LPTouchUtils applicationWindows];
+  NSArray *results = [parser evalWith:allWindows];
+  if (!results || results.count == 0) {
+     return nil;
+  } else {
+     return results[0];
+  }
+}
+
+// post_key_delay =>
+// string =>
+- (NSDictionary *) JSONResponseForMethod:(NSString *) method
+                                     URI:(NSString *) path
+                                    data:(NSDictionary *) data {
+
+  LPLogDebug(@"data: %@", data);
+
+  if (data[@"post_key_delay"]) {
+    self.postKeyDelay = (CFTimeInterval)[data[@"post_key_delay"] floatValue];
+  }
+
+  LPLogDebug(@"LPTypeString post key delay = %@", @(self.postKeyDelay));
+
+  NSString *reason = @"";
+
+  id target = [self firstResponder];
+  if (!target) {
+    reason = [NSString stringWithFormat:@"Cannot type text because no view "
+                                            " is first responder with query: '%@'",
+                                        LPTypeStringFirstResponderQuery];
+    LPLogError(@"%@", reason);
+    return [self failureResponseWithReason:reason];
+  }
+
+  NSString *textToType = data[@"string"];
+  if (!textToType) {
+    reason = [NSString stringWithFormat:@"Request is missing value for 'string' key"];
+    LPLogError(@"%@", reason);
+    return [self failureResponseWithReason:reason];
+  }
+
+
+  if (![target conformsToProtocol:@protocol(UITextInputTraits)]) {
```

Comment:
UITextField, UITextView, and UISearchBar all conform to this protocol.

It does look like a bug because id<UIKeyInput> might not conform to this protocol.